### PR TITLE
[DI] Fix merging of env vars in configs

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 3.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #22561, #23520
| License       | MIT
| Doc PR        | -

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
Because processing configuration is done internally in `Extension::load()`, we have no proper way to get the processed configuration from the outside.

I didn't find any better implementation than the one in this PR, which consist of storing processed configs when `Extension::processConfiguration()` is being called, then fetching these with a new `Extension::getProcessedConfigs()` method.

This works only if extensions do call `processConfiguration()` of course.
I'd expect this to be a common enough situation to consider the issue fixed.
If the expectation is not met, the outcome is that overridden env vars are tracked again, which is not a huge issue anyway, and is just the current situation.

Decoupling config processing from loading is another possibility, but that'd be asking for big changes for all bundle authors. Not worth it IMHO.

PR is ready (failures are false positives.)

File: src/Symfony/Component/DependencyInjection/Extension/Extension.php [hunk 1]
```diff
@@ -91,7 +93,16 @@ public function getConfiguration(array $config, ContainerBuilder $container)
     {
         $processor = new Processor();
 
-        return $processor->processConfiguration($configuration, $configs);
+        return $this->processedConfigs[] = $processor->processConfiguration($configuration, $configs);
+    }
+
+    final public function getProcessedConfigs()
```
Comment (fabpot) [inline src/Symfony/Component/DependencyInjection/Extension/Extension.php]: /** @internal */?

**Review by fabpot — CHANGES_REQUESTED** [review]
I think we need copious comments so that we can still understand the logic in 2 years :)

File: src/Symfony/Component/DependencyInjection/Extension/Extension.php (hunk 1 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Extension/Extension.php]: so it is

**Comment by nicolas-grekas** [comment]
comments added, should be ready

File: src/Symfony/Component/DependencyInjection/ParameterBag/EnvPlaceholderParameterBag.php [hunk 2]
```diff
@@ -106,11 +106,11 @@ public function resolve()
     /**
      * Replaces "%env(FOO)%" references by their placeholder, keeping regular "%parameters%" references as is.
      */
-    public function resolveEnvReferences()
+    public function resolveEnvReferences(array $value)
```
Comment (Taluu) [inline src/Symfony/Component/DependencyInjection/ParameterBag/EnvPlaceholderParameterBag.php]: isn't this a BC break ?

Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/ParameterBag/EnvPlaceholderParameterBag.php]: this is not released yet, 3.3@dev only

**Comment by nicolas-grekas** [comment]
Merging to unlock another PR to come fixing a bug in 3.3 also.
If you have further comments please still add them.

**Pull request merged**